### ISSUE TITLE
TARGETID inconsistent with BRICKID, BRICK_OBJID

### ISSUE THREAD

**Issue by sbailey** [desihub/fiberassign#423]
Fiberassign files in the range 4-596 and 80615-81115 have some targets with BRICKID, BRICK_OBJID that is inconsistent with those encoded in TARGETID, as decoded with `desitarget.targets.decode_targetid(targetid)`.  Most but not all of these are secondary targets and I suspect are a leftover of issue #403 or #375.  Curiously some have (SV3_)DESI_TARGET=0 which also seems problematic (a target with no targeting bits?)

e.g.
```
from astropy.table import Table
from desitarget.targets import decode_targetid
fa = Table.read('/global/cfs/cdirs/desi/target/fiberassign/tiles/trunk/000/fiberassign-000010.fits.gz', 'FIBERASSIGN')
objid, brickid = decode_targetid(fa['TARGETID'])[0:2]

bad = (fa['TARGETID']>0) & ((fa['BRICKID'] != brickid) | (fa['BRICK_OBJID'] != objid))
fa['TARGETID', 'SV3_DESI_TARGET', 'SV3_SCND_TARGET', 'BRICKID', 'BRICK_OBJID'][bad]

-->
     TARGETID       SV3_DESI_TARGET   SV3_SCND_TARGET BRICKID BRICK_OBJID
      int64              int64             int64       int32     int32   
----------------- ------------------- --------------- ------- -----------
39627817444442354 4611686018427387904     34359738368       0           0
 1083331777134598 4611686018427387904  70368744177664       0           0
39627811400452830 4611686018427387904       100663296       0           0
 1017355039670272 4611686018427387904      2147483648       0           0
39627811400451183 4611686018427387904           65536       0           0
43977491483722165                   0               0       0           0
39627817440249125 4611686018427387904          131072       0           0
43977491479527857                   0               0       0           0
...
```

```
extrabad = bad & (fa['BRICKID'] != 0)
fa['TARGETID', 'SV3_DESI_TARGET', 'SV3_SCND_TARGET', 'BRICKID', 'BRICK_OBJID'][extrabad]
     TARGETID       SV3_DESI_TARGET   SV3_SCND_TARGET BRICKID BRICK_OBJID
      int64              int64             int64       int32     int32   
----------------- ------------------- --------------- ------- -----------
39627817431859705 4611686018427387904       100663296  335285        5332
39627829515652507 4611686018427387904       100663296  339604        1610
39627823467468138 4611686018427387904       100663296  338164          93
39627817436056009 4611686018427387904     34359738368  335288        4787
39627841586857297 4611686018427387904       100663296  341052        4643
39627847618265806 4611686018427387904       100663296  342492         117
39627841586857764 4611686018427387904       100663296  341053         170
39627823488437387 4611686018427387904       100663296  338167        2725
```

@dstndstn @araichoor @geordie666 can you track these down, identify the root cause of the mismatch (induced by fiberassign, or something upstream in the targeting), and see what it would take to patch these in fiberassign svn?  (I recognize that @geordie666 usually recommends documenting over patching, but let's at least scope that out).  @moustakas's work matching targets back to tractor files in desihub/desispec#1716 may help too.

FWIW, commit afceb03 has a fiberassign/bin/patch-fa.py that has since been removed.  It looks like it considered mismatches of BRICKID and BRICK_OBJID, but then didn't actually patch them (RELEASE too).  That script appears to be "here's what I did for the record..." but not necessarily a general purpose tool for the future (so be cautious about reusing it).

**Comment by geordie666** [comment]
For context:

Anything which has `SV3_DESI_TARGET==4611686018427387904` (_just_ the `SCND_ANY` bit of `2**62` set) is a "secondary only" target, or what I like to call a "standalone secondary". The other TARGETIDs (`43977491483722165`, etc.) look like they all have `RELEASE=9999`, meaning they are Targets of Opportunity. 

For all of these targets, any `BRICKID` and `BRICK_OBJID` columns that people choose to add downstream of `desitarget` should always be zero. These columns (deliberately) _do not exist_ in the corresponding secondary targeting files — because the `BRICKID` and `BRICK_OBJID` have a distinct meaning for standalone secondaries and ToOs. This is because standalone secondaries and ToOs _might not be drawn from DR9 of the Legacy Surveys_. 

If you're interested in further detail, how `TARGETID` is defined for such sources is in Section 3.2 of the [desitarget paper](https://desi.lbl.gov/DocDB/cgi-bin/private/ShowDocument?docid=6693).

My hope is that this is, indeed, the fiberassign bug. I personally don't believe we should be populating `BRICKID` or `BRICK_OBJID` for any of these types of targets downstream of `desitarget`.

**Comment by araichoor** [comment]
Just trying to understand / clarify the situation, I ve digged a bit more in the example SV3 fiberassign-000010.fits.gz file.
There are 335 identifed `bad` rows in that example: 309 standalone secondary and 26 ToOs.

Standalone secondary targets:
```
scnd = Table.read("/global/cfs/cdirs/desi/target/catalogs/dr9/0.57.0/targets/sv3/secondary/dark/sv3targets-dark-secondary.fits")
scndbad = (bad) & (np.in1d(fa["TARGETID"], scnd["TARGETID"]))
scndbad.sum(), ((scndbad) & (fa["SV3_DESI_TARGET"] != 2 ** 62)).sum()
=>
(309, 0)
```
If I understand @geordie666 message, the `decode_targetid` should not be used on those, so it s not a problem to have `fa["BRICKID"] != brickid`.
But 8 of those have `BRICKID != 0` (the `extrabad` ones): here, it sounds like the fiberassign bug with rows mismatching when merging the various inputs to create the fiberassign-TILEID.fits.gz

ToOs:
Same, it s not a problem to have `fa["BRICKID"] != brickid`.
I notice that `decode_targetid ` identifies for those 26 ToOs `brickid!=9999`:
```
too = Table.read("/global/cfs/cdirs/desi/survey/fiberassign/SV3/20210503/000010-too.fits")
toobad = (bad) & (np.in1d(fa["TARGETID"], too["TARGETID"]))
toobad.sum()
np.unique(brickid[toobad]).tolist()
```
=>
26
[336725, 336730, 338164, 338170, 339607, 339608, 339609, 341044, 341045, 341046, 341047, 342484, 342485, 342486, 342487, 343926]

**Comment by geordie666** [comment]
One clarification: Although I agree in both cases that it's not a problem to have `fa["BRICKID"] != brickid` I would certainly prefer if that was because `fa["BRICKID"] == 0` (or some other null value).

`decode_targetid()` _can_ and _should_ be used in those cases. The resulting values of:
```
o, b = decode_targetid(fa['TARGETID'])[0:2]
``` 
Are _completely meaningful_. The `o` and `b` just shouldn't be construed to correspond to `BRICKID` or `OBJID` from the Legacy Surveys for standalone secondary targets or ToO targets.

**Comment by araichoor** [comment]
doing my homework, I m re-reading all the fiberassign threads related to those bug issue.

for the record, about the sentence from @sbailey in the original message: "_Curiously some have (SV3)DESI_TARGET=0 which also seems problematic (a target with no targeting bits?)_"
=>
this is indeed another identified issue.
in the early days when we added ToOs, no ToO columns were propagated to the `fiberassign-TILEID.fits.gz` (including the `SV3_*TARGET`), hence filled with zeros.

from a code point-of-view this has been fixed in https://github.com/desihub/fiberassign/pull/342 (included in fiberassign/2.5.0).
but sv3 tiles designed before that are not fixed yet.

it will be fixed with the second round of patching.

**Comment by araichoor** [comment]
This is fixed in https://github.com/desihub/fiberassign/pull/436, used to do the second patching the fiberassign files.
Those patched fiberassign files will shortly be committed to svn, and will be used by future desispec productions (himalayas? iron, etc).
